delaunay: implement delaunay triangulation

Bastiantheone opened the pull request.

Bastiantheone commented:
@sbinet can you help me make this pull request pass the checks?

sbinet commented:
done :)
(the travis builds have a bit of "lstat flakiness"...)

Bastiantheone commented:
Great :) @sbinet to combine the two commits into one: `git rebase -i ` what do I need to put after the `-i`? origin/Delaunay?

sbinet commented:
Or 'origin/master'.

Bastiantheone commented:
It's rebased

sbinet commented:
thanks.
apologies for the belated reply.
I'll look into this ASAP (but I am at a conference the whole week, so I'll lag a bit...)

sbinet commented on fastjet/internal/delaunay/plot.go:
```diff
@@ -0,0 +1,70 @@
+		}
+		p.Add(poly)
+	}
+	p.Save(1000, 1000, path)
```
add a test of the error or return that error.

sbinet commented on fastjet/internal/delaunay/point.go:
```diff
@@ -0,0 +1,99 @@
+package delaunay
```
please add a license header

sbinet commented on fastjet/internal/delaunay/point.go:
```diff
@@ -0,0 +1,99 @@
+	"strconv"
+)
+
+type Point struct {
```
Please document that type as well as the various fields of that struct.

sbinet commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,201 @@
+package delaunay
```
please add a license header

sbinet commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,201 @@
+	"math"
+)
+
+type Triangle struct {
```
please document that type (and its fields)

sbinet commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,201 @@
+func NewTriangle(a *Point, b *Point, c *Point) *Triangle {
+	// make it counter clockwise
+	if !cClockwise(a, b, c) {
+		temp := b
```
that could be re-written as:
```go
b, c = c, b
```

sbinet commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,201 @@
+	// r is the radius of the circumcircle
+	r := math.Sqrt((x-t.A.x)*(x-t.A.x) + (y-t.A.y)*(y-t.A.y))
+	// point is in circumcircle if distance to center is less than radius
+	return math.Sqrt((p.x-x)*(p.x-x)+(p.y-y)*(p.y-y)) < r
```
why not just compare the squares instead of the square roots?
that should give the same answer w/o involving the quite costly call to `math.Sqrt`, shouldn't it ?

Bastiantheone replied:
Since it is very expensive does it make sense to keep the distances squared internally since it is used only for comparisons? And add a method to get the distance for the user, where the square root is taken.

sbinet commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,201 @@
+func (t *Triangle) centCCircle() (float64, float64) {
+	m1 := (t.A.x - t.B.x) / (t.B.y - t.A.y)
+	m2 := (t.A.x - t.C.x) / (t.C.y - t.A.y)
+	b1 := (t.A.y+t.B.y)/2 - m1*(t.A.x+t.B.x)/2
```
I would multiply by `0.5` instead of dividing by `2`

sbinet commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,646 @@
+package delaunay
```
please add a license header

sbinet commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,646 @@
+	"fmt"
+)
+
+type Delaunay struct {
```
please document this type as well as its fields.

sbinet commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,331 @@
+package delaunay
```
please add a license header

sbinet commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,331 @@
+	"math"
+)
+
+type Voronoi struct {
```
please document this type and its fields.

sbinet commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,331 @@
+		}
+		m = (y2 - y1) / (x2 - x1)
+		if math.IsNaN(m) {
+			// FIXME
```
add a temporary panic so we don't forget ?
(also probably add an issue)

sbinet commented:
as a very high-level comment about the overall layout of the code: I am bit worried about all the pointers to `*Point` and `*Triangle`.
I see you have devised a few benchmark functions.
would it be possible (and easy on the client API side) to have more values throughout? (ie: copy the `Point`s and `Triangle` ?) and check the impact on the benchmarks ?

alternatively, could you also publish some benchmark numbers (CPU+memalloc) together with the PR?

Bastiantheone commented:
Here are test data of the current version before any changes: 

_(pprof) top20
33960ms of 37220ms total (91.24%)
Dropped 114 nodes (cum <= 186.10ms)
Showing top 20 nodes out of 74 (cum >= 390ms)
      flat  flat%   sum%        cum   cum%
    7330ms 19.69% 19.69%     7330ms 19.69%  delaunay.(*Triangle).Equals
    3680ms  9.89% 29.58%     3680ms  9.89%  math.Sqrt
    2970ms  7.98% 37.56%     6440ms 17.30%  delaunay.(*Point).distance
    2940ms  7.90% 45.46%    21680ms 58.25%  delaunay.(*Delaunay).validateEdge
    2830ms  7.60% 53.06%    12250ms 32.91%  delaunay.remove
    2260ms  6.07% 59.13%     5530ms 14.86%  runtime.mallocgc
    2130ms  5.72% 64.86%     2130ms  5.72%  runtime.heapBitsSetType
    1880ms  5.05% 69.91%     6290ms 16.90%  delaunay.(*Point).nearestNeighbor
    1250ms  3.36% 73.27%     4410ms 11.85%  delaunay.appendT
    1070ms  2.87% 76.14%     2330ms  6.26%  delaunay.NewVoronoi
     890ms  2.39% 78.53%     1480ms  3.98%  runtime.scanobject
     800ms  2.15% 80.68%      920ms  2.47%  runtime.nextFreeFast
     620ms  1.67% 82.35%      620ms  1.67%  delaunay.(*Triangle).centCCircle
     610ms  1.64% 83.99%      880ms  2.36%  runtime.typedslicecopy
     520ms  1.40% 85.38%     2770ms  7.44%  runtime.growslice
     510ms  1.37% 86.75%      510ms  1.37%  delaunay.(*Triangle).inTriangle
     480ms  1.29% 88.04%    12300ms 33.05%  delaunay.(*Delaunay).flip
     410ms  1.10% 89.15%    26860ms 72.17%  delaunay.(*Delaunay).insertP
     390ms  1.05% 90.19%      390ms  1.05%  runtime.heapBitsForObject
     390ms  1.05% 91.24%      390ms  1.05%  runtime.memmove
(pprof) top --cum
3.85s of 37.22s total (10.34%)
Dropped 114 nodes (cum <= 0.19s)
Showing top 10 nodes out of 74 (cum >= 21.68s)
      flat  flat%   sum%        cum   cum%
         0     0%     0%     34.43s 92.50%  runtime.goexit
         0     0%     0%     34.41s 92.45%  testing.(*B).runN
         0     0%     0%     32.20s 86.51%  testing.(*B).launch
         0     0%     0%     31.85s 85.57%  delaunay_test.benchmarkDelaunay
     0.02s 0.054% 0.054%     27.81s 74.72%  delaunay.NewDelaunay
     0.04s  0.11%  0.16%     27.59s 74.13%  delaunay.(*Delaunay).InsertPoint
     0.41s  1.10%  1.26%     26.86s 72.17%  delaunay.(*Delaunay).insertP
     0.24s  0.64%  1.91%     24.08s 64.70%  delaunay.(*Delaunay).RemovePoint
     0.20s  0.54%  2.44%     21.70s 58.30%  delaunay.(*Delaunay).removeP
     2.94s  7.90% 10.34%     21.68s 58.25%  delaunay.(*Delaunay).validateEdge_

And bechmark results:
Nr of points, execution time
_50,	   8700513
100,	  19871151 
150,	  30101770 
![scatter-plot-execution-time](https://user-images.githubusercontent.com/23250770/27838409-722b4f3e-60b8-11e7-9e71-f5a8c3b342d5.png)

200,	  40935753 
250,	  50303033 
300,	  63603620 
350,	  75904350 
400,	  85454995 
450,	  96205630 
500,	 109606110 
550,	 127505020 
600,	 134707680 
650,	 145606530 
700,	 154308390 
750,	 167009780 
800,	 179010610 
850,	 191311060 
900,	 208011900 
950,	 218016700 
1000, 228613220_

Bastiantheone commented:
Your suggestions have been added. Should I wait to add the issues until it's merged or can I add them now?

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,206 @@
+	return t
+}
+
+func cClockwise(a *Point, b *Point, c *Point) bool {
```
Please rename this `clockwise` and invert the logic throughout.

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,206 @@
+		B: b,
+		C: c,
+	}
+	t.children = make([]*Triangle, 0, 10)
```
Don't bother with this.

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,206 @@
+	return (b.Y-a.Y)*(c.X-b.X)-(c.Y-b.Y)*(b.X-a.X) < 0
+}
+
+// checking if the point is in the triangle and if it is on an edge
```
"// inTriangle checks whether the point ..."

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,206 @@
+
+// checking if the point is in the triangle and if it is on an edge
+// using barycentric coordinates
+func (t *Triangle) inTriangle(p *Point) (bool, bool) {
```
Please used named returns on the booleans, `(inside, edge bool)`

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,206 @@
+		return true, false
+	}
+	// outside triangle
+	if barcen1 < 0 || barcen2 < 0 || barcen3 < 0 {
```
```
outside := barcen1 < 0 || barcen2 < 0 || barcen3 < 0
return !outside, !outside
```

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,206 @@
+	return (t.B.Y-t.C.Y)*(t.A.X-t.C.X) + (t.C.X-t.B.X)*(t.A.Y-t.C.Y)
+}
+
+// return true if the point is inside the circumcircle of the triangle
```
"// inCircumCircle returns whether the point is inside the circumcircle of the triangle."

and rename the method.

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,206 @@
+}
+
+// returns the center of the circumcircle
+func (t *Triangle) centCCircle() (float64, float64) {
```
```
// centerOfCircumCircle returns the the triangle's circum circle.
func (t *Triangle) centerOfCircumCircle() (x, y float64) {
```
and change the `x` and `y` decls below.

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,206 @@
+	return ft
+}
+
+func (t1 *Triangle) Equals(t2 *Triangle) bool {
```
`s/t1/t/g` (and `s/t2/other/g`?)

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,206 @@
+	if t1 == t2 {
+		return true
+	}
+	if !t1.A.Equals(t2.A) && !t1.A.Equals(t2.B) && !t1.A.Equals(t2.C) ||
```
This can be simplified to not use a conditional.

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,206 @@
+	return true
+}
+
+func (t *Triangle) ToString(name string) string {
```
```
func(t *Triangle) String() string {
    return fmt.Sprintf("{A%s, B%s, C%s}", t.A, t.B, t.C)
}
```

or something similar (braces help clarify output here).

Bastiantheone commented on fastjet/internal/delaunay/delaunay_test.go:
```diff
@@ -0,0 +1,293 @@
+const tol = 1e-3
+
+func TestSimple(t *testing.T) {
+	a := delaunay.NewPoint(0, 0, 0)
```
I feel like this would make it very hard to follow when it comes to removing points and checking the results. Do you think I can just leave a comment `// NewPoint(x,y,id)`

kortschak replied:
It was more for the simplification. I was ambivalent, because of the later operations you do, so you can just leave this.

Bastiantheone commented on fastjet/internal/delaunay/point.go:
```diff
@@ -0,0 +1,104 @@
+	Y       float64     // Y coordinate
+	adjT    []*Triangle // holds all triangles adjacent to a point
+	outside bool        // true if point is on the outside of the triangulation
+	Ind     int         // user defined identifier
```
No, it doesn't have any requirements. I will use it for the clustering in fastjet and I'm using it to remove points.

kortschak replied:
Can't you use the pointer value for that and remove the field completely?

Bastiantheone replied:
I think it might be possible, but it would make things way more complicated. Especially when it comes to corresponding jets to points. Do you think the ID field needs to be removed?

kortschak replied:
If it needs to be there, then it needs to be there, but if it doesn't it shouldn't be.

From the way you describe its use, with no invariants, then a pointer value would do, but I see in `inPolygon` there clearly are invariants that are not supported by the use of pointer comparison; it seems to me from reading that that IDs must be unique within the set of points - this would work using pointer values, but it is not safe since the pointer values may change under foot (and pointer inequality is not completely defined). You should make this clear in the documentation.

Can you outline how you plan to use this in your problem so I can think about it more in context.

Bastiantheone replied:
I need the ID for two things: 

To remove points: When a point is removed copies of the adjacent points are made and a new triangulation with just those copies is started. After it ran through, the ID tells me which copy belongs to which point and the point can be updated accordingly. 

For the jet clustering: I will need to know which jet corresponds to which point

kortschak replied:
OK, so you do need IDs here. I am sort of at a loss how they work for you in `inPolygon`. Can you explain that?

kortschak replied:
OK, I have that sorted out.

Rename `inPolygon` to `areClockwise` and then make it clear that triangles have points with IDs that exist such that `t.A.ID < t.B.ID < t.C.ID`. I don't see how that means what it says otherwise.

Bastiantheone replied:
I apologize, I forgot to mention that I need it in inPolygon. And yes that is mostly true. The points on the polygon are indexed in order. If there are 4 points around the point to be removed, they are indexed counterclockwise. Triangle with index 4,1,2 is inside the polygon, just like 1,2,3 or 3,4,1 for example. But Triangle 1,3,2 would be outside the polygon. That works, because A,B,C are ordered counterclockwise

Bastiantheone replied:
![20170706_223601](https://user-images.githubusercontent.com/23250770/27940943-dd2bc87e-629b-11e7-8744-b32ba84989c8.jpg)
I hope this will make a llittle more sense

Bastiantheone commented:
Here are benchmark results for insertion and removal. The number of points increases each line by 50. Is this what you are looking for?
```
200	   8450404 ns/op	  167588 B/op	    7149 allocs/op
100	  19951152 ns/op	  371418 B/op	   15796 allocs/op
50	  30801874 ns/op	  567222 B/op	   24319 allocs/op
30	  42369080 ns/op	  754806 B/op	   32177 allocs/op
30	  54670566 ns/op	  990789 B/op	   41342 allocs/op
20	  66653920 ns/op	 1194115 B/op	   50034 allocs/op
20	  77804665 ns/op	 1365716 B/op	   57871 allocs/op
20	  88005060 ns/op	 1593446 B/op	   66640 allocs/op
20	  94755520 ns/op	 1717412 B/op	   72162 allocs/op
10	 112406060 ns/op	 2025206 B/op	   84487 allocs/op
10	 119808770 ns/op	 2186985 B/op	   91407 allocs/op
10	 133907800 ns/op	 2418468 B/op	  101110 allocs/op
10	 146410500 ns/op	 2652504 B/op	  110026 allocs/op
10	 158309150 ns/op	 2810996 B/op	  117493 allocs/op
10	 170909720 ns/op	 3047400 B/op	  126786 allocs/op
10	 177410510 ns/op	 3168212 B/op	  132310 allocs/op
10	 193609330 ns/op	 3435756 B/op	  142744 allocs/op
10	 204311950 ns/op	 3622916 B/op	  150726 allocs/op
5	 215412200 ns/op	 3830464 B/op	  159513 allocs/op
5	 226213540 ns/op	 4021262 B/op	  167995 allocs/op
```

kortschak commented:
By way of advice, for a fairly involved PR such as this, I would not squash until the end. Re-reading during review makes is harder for reviewers to be sure that comments have been addressed. Also, replying "Done" or similar to comments also helps.

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+}
+
+// if points are ordered clockwise they are outside the polygon
+func inPolygon(a, b, c int) bool {
```
```
if b < c {
    return a < b || c < a
} else {
    return a < b && c < a
}
```

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,193 @@
+
+// inTriangle checks whether the point is in the triangle and whether it is on an edge
+// using barycentric coordinates
+func (t *Triangle) inTriangle(p *Point) (inside, edge bool) {
```
This could benefit from inversion of the signature. It reads more naturally as a method on `*Point`, `func (p *Point) inTriangle(t *Triangle) (inside, edge bool)`.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,193 @@
+}
+
+// inCircumCircle returns whether the point is inside the circumcircle of the triangle
+func (t *Triangle) inCircumCircle(p *Point) bool {
```
Sorry, I got the capitalisation wrong here it should be `inCircumcircle`.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,193 @@
+}
+
+// centerOfCircumCircle returns the center of the triangle's circum circle.
+func (t *Triangle) centerOfCircumCircle() (x, y float64) {
```
Same alteration here `centerOfCircumcircle`.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,193 @@
+}
+
+// appends and updates nearest neighbor
+func appendT(triangles []*Triangle, elems ...*Triangle) []*Triangle {
```
If you make a `type triangles []*Triangle` you get `(*triangles).append`, `(*triangles).remove` and `triangles.finalize` as namespaced methods which improves clarity. Then just change the `children` field to `triangles`.

Is this function used somewhere?

Bastiantheone replied:
Yes, appendT is used when the adjacent Triangles of a point are updated. It checks if the nearest neighbor changed and updates it accordingly. Then it appends the new triangles to the adjacent triangles list.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,193 @@
+// removes given triangles from a slice of triangles
+func remove(triangles []*Triangle, elems ...*Triangle) []*Triangle {
+	// check if nearest neighbor of any point is removed
+	update := make([]*Point, 0)
```
`var update []*Point`

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,193 @@
+	// x and y are the coordinates of the center of the circumcircle
+	x = (b2 - b1) / (m1 - m2)
+	y = m1*x + b1
+	if math.IsNaN(x) || math.IsNaN(y) {
```
If x is NaN, then y is by propagation, so you only need to check y.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,193 @@
+		b2 = (t.B.Y+t.C.Y)*0.5 - m2*(t.B.X+t.C.X)*0.5
+		x = (b2 - b1) / (m1 - m2)
+		y = m1*x + b1
+		if math.IsNaN(x) || math.IsNaN(y) {
```
Same here.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,193 @@
+			b2 = (t.C.Y+t.B.Y)*0.5 - m2*(t.C.X+t.B.X)*0.5
+			x = (b2 - b1) / (m1 - m2)
+			y = m1*x + b1
+			if math.IsNaN(x) || math.IsNaN(y) {
```
And here.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+
+// FIXME can't do any delaunay operation after calling NewVoronoi
+func NewVoronoi(d *Delaunay) *Voronoi {
+	border := make([]*Triangle, 0, len(d.root.A.adjacentTriangles)+len(d.root.B.adjacentTriangles)+len(d.root.C.adjacentTriangles))
```
Since you've allocated `border` with known size, you can use `copy` here rather than `append`.

Bastiantheone replied:
Are you suggesting to copy the first adjacent triangle `copy(border,d.root.A.adjacentTriangles)`, and the rest is appended or all of them? If so how would that be implemented?

Bastiantheone replied:
If not it's done

kortschak replied:
```
n := copy(border, d.root.A.adjacentTriangles)
n += copy(border[n:], d.root.B.adjacentTriangles)
copy(border[n:], d.root.C.adjacentTriangles)
```

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+// FIXME can't do any delaunay operation after calling NewVoronoi
+func NewVoronoi(d *Delaunay) *Voronoi {
+	border := make([]*Triangle, 0, len(d.root.A.adjacentTriangles)+len(d.root.B.adjacentTriangles)+len(d.root.C.adjacentTriangles))
+	border = append(border, d.root.A.adjacentTriangles[:]...)
```
The `[:]` in these lines is not necessary, `adjacentTriangles` is already a `[]*Triangle` (or `triangles`).

Bastiantheone replied:
Same as above

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+	border = append(border, d.root.B.adjacentTriangles[:]...)
+	border = append(border, d.root.C.adjacentTriangles[:]...)
+	for i, t := range d.triangles {
+		if len(d.triangles[i].children) == 0 {
```
`s/if len(d.triangles[i].children) == 0 {/if len(t) == 0 {/`, or preferably

```
if len(t) != 0 {
    continue
}
```

to reduce the indentation in the work block.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+
+// Voronoi holds the border information
+type Voronoi struct {
+	maxX float64
```
These can all be on one line as `maxX, minX, maxY, minY float64`.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+		}
+	}
+	v := &Voronoi{maxX: d.maxX, minX: d.minX, maxY: d.maxY, minY: d.minY}
+	return v
```
`return &Voronoi{maxX: d.maxX, minX: d.minX, maxY: d.maxY, minY: d.minY}`

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+
+// finds the voronoi area of a point
+// returns the area and the points associated with it
+func (v *Voronoi) VoronoiCell(p *Point) (float64, []*Point) {
```
Use named returns here, `(area float64, centers []*Point)`.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+		// find first triangle on the outside
+		for _, t1 := range p.adjacentTriangles {
+			var pt *Point
+			if t1.A.Equals(p) {
```
This reads better as a switch (the same code will be generated).

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+		}
+		first = t
+		if t == nil {
+			// should only happen when user makes a mistake
```
What kind of mistake? The panic message should be more informative, so they can understand what they did. Then remove the comment, since the code will explain itself.

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+		// FIXME problem when circumcenter is outside border
+		// extend polygon to border
+		var pt, pt2 *Point
+		if p.Equals(last.A) {
```
`switch`

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+			pt2 = last.B
+		} else {
+			// should never happen
+			panic(fmt.Errorf("voronoi: internal error with adjacent triangles for point P%s", p))
```
Make the code comment itself.

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+		}
+		m := (y2 - y1) / (x2 - x1)
+		if math.IsNaN(m) {
+			// FIXME
```
Please say what is needing fixing.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+			panic(fmt.Errorf("delaunay: not implemented for the constellation in triangle T%s", last))
+		}
+		var side int
+		const top, right, bottom, left = 0, 1, 2, 3
```
```
const (
    top = iota + 1
    right
    bottom
    left
)
```

Avoiding the zero allows failure to be detected.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+		}
+		var side int
+		const top, right, bottom, left = 0, 1, 2, 3
+		if x2-x1 > 0 { // right
```
`if x2 > x1 {`, similarly for below.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+			}
+		}
+		// find second border point
+		if p.Equals(first.A) {
```
`switch`

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+			pt2 = first.A
+		} else {
+			// should never happen
+			panic(fmt.Errorf("voronoi: internal error with adjacent triangles for point P%s", p))
```
Comment in panic rather than in comment.

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+		}
+		m = (y2 - y1) / (x2 - x1)
+		if math.IsNaN(m) {
+			// FIXME
```
State problem.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+			last = t
+		}
+	}
+	if p.isOutside {
```
If this condition is inverted, you get

```
if !p.isOutside {
    return polyArea(centers), centers
}
```

and the resulting else block can be deindented one level.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+// finds the area of an irregular polygon
+// needs the points in clockwise order
+func polyArea(points []*Point) float64 {
+	area := 0.0
```
`var area float64`

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+		area += (points[j].X + points[i].X) * (points[j].Y - points[i].Y)
+		j = i
+	}
+	return area / 2
```
@sbinet will want this to be `0.5 * area`.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+func findClockTri(p *Point, t *Triangle) *Triangle {
+	// points in a triangle are ordered counter clockwise
+	var p2 *Point
+	if p.Equals(t.A) {
```
`switch`

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,343 @@
+	} else if p.Equals(t.C) {
+		p2 = t.A
+	} else {
+		// should only happen when user makes a mistake
```
Comment is unnecessary.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+type Delaunay struct {
+	triangles []*Triangle // all triangles created
+	root      *Triangle   // triangle that contains all points. Used to find a triangle
+	maxX      float64
```
Single line.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+func (d *Delaunay) GetTriangles() []*Triangle {
+	// remove triangles that contain the root points
+	rt := make([]*Triangle, 0, len(d.root.A.adjacentTriangles)+len(d.root.B.adjacentTriangles)+len(d.root.C.adjacentTriangles))
+	rt = append(rt, d.root.A.adjacentTriangles[:]...)
```
Use `copy` rather than `append`, `[:]` is not necessary.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+	return d
+}
+
+func (d *Delaunay) GetTriangles() []*Triangle {
```
`func (d *Delaunay) Triangles() []*Triangle {`; the Go convention is that getters are named for the thing they get, setters are named `Set<Thing>`.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+// holds necessary information for the
+// delaunay triangulation
+type Delaunay struct {
+	triangles []*Triangle // all triangles created
```
`triangles triangles`

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+	rt = append(rt, d.root.A.adjacentTriangles[:]...)
+	rt = append(rt, d.root.B.adjacentTriangles[:]...)
+	rt = append(rt, d.root.C.adjacentTriangles[:]...)
+	return finalize(d.triangles, rt[:]...)
```
`return d.triangles.finalize(rt...)`

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+}
+
+func (d *Delaunay) InsertPoint(p *Point) {
+	p.adjacentTriangles = make([]*Triangle, 0)
```
This is not necessary.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+	}
+	// remove triangles adjacent to p
+	for _, t := range p.adjacentTriangles {
+		if t.A.Equals(p) {
```
`switch`

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+	// if that's the case either an error has occurred or the point is outside or on the root triangle
+	if len(p.adjacentTriangles) < 3 {
+		// should only happen when user makes a mistake
+		panic(fmt.Errorf("delaunay: can't remove point P%s", p))
```
`panic(fmt.Errorf("delaunay: can't remove point P%s not enough triangles", p))` or something similar.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+func (d *Delaunay) RemovePoint(p *Point) {
+	// if that's the case either an error has occurred or the point is outside or on the root triangle
+	if len(p.adjacentTriangles) < 3 {
+		// should only happen when user makes a mistake
```
Delete comment.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+}
+
+func (d *Delaunay) RemovePoint(p *Point) {
+	// if that's the case either an error has occurred or the point is outside or on the root triangle
```
This comment is unclear; what is "that".

Bastiantheone replied:
That was referring to the if statement, but I removed the comment

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+	t := p.adjacentTriangles[0]
+	j := 1
+	k := 0
+	if t.A.Equals(p) {
```
`switch`

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+		points[1] = t.B
+	} else {
+		// should never happen
+		panic(fmt.Errorf("delaunay: internal error with adjacent triangles for P%s and T%s", p, t))
```
Clarify panic and remove comment.

Bastiantheone replied:
The panic should never happen because a point should always be a point in all triangles in its adjacentTriangles list. How should I word the panic?

kortschak replied:
`panic(fmt.Errorf("delaunay: point %v not in adjacent triangle %v", p, t))`

BTW, this could just be fmt.Sprintf, unless you are wanting people to recover and type switch on the recovered value. Not a big issue, but something to consider in the future.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+	for i := 0; j < len(points)-1; {
+		if i >= len(p.adjacentTriangles) {
+			// should only happen when user makes a mistake
+			panic(fmt.Errorf("delaunay: internal error with adjacent triangles for P%s", p))
```
Here too.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+			continue
+		}
+		t = p.adjacentTriangles[i]
+		if t.A.Equals(points[j]) {
```
`switch`

Bastiantheone replied:
done

kortschak commented:
Thank. So there should be some explanation of that invariant in the
field doc comments.

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+	if !inside {
+		return nil, false
+	}
+	if len(t.children) == 0 {
```
```
if len(t.children) == 0 {
    return t, edge
}
```

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+		}
+	}
+	// find points opposite and adjacent to the edge
+	var p1 *Point
```
These can all be on one line.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+	var p2 *Point
+	var pO1 *Point
+	var pO2 *Point
+	if !t1.A.Equals(t2.A) && !t1.A.Equals(t2.B) && !t1.A.Equals(t2.C) {
```
`switch`

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+		pO1 = t1.B
+		pO2 = t1.A
+	}
+	if !t2.A.Equals(t1.A) && !t2.A.Equals(t1.B) && !t2.A.Equals(t1.C) {
```
`switch`

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+// turns triangle into a valid delaunay triangle
+func (d *Delaunay) validateEdge(t *Triangle, p *Point) {
+	// find points in the triangle that are not p
+	var p2 *Point
```
One line.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+	// find points in the triangle that are not p
+	var p2 *Point
+	var p3 *Point
+	if t.A.Equals(p) {
```
`switch`

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+		p2 = t.A
+		p3 = t.B
+	} else {
+		// should never happen
```
Remove comment.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+// flip edge between two triangles
+func (d *Delaunay) flip(t1 *Triangle, t2 *Triangle) (*Triangle, *Triangle) {
+	// find points adjacent and opposite to edge
+	var p1 *Point
```
One line.

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+	var p2 *Point
+	var pO1 *Point
+	var pO2 *Point
+	if !t1.A.Equals(t2.A) && !t1.A.Equals(t2.B) && !t1.A.Equals(t2.C) {
```
`switch`

Bastiantheone replied:
done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+		pO1 = t1.B
+		pO2 = t1.A
+	}
+	if !t2.A.Equals(t1.A) && !t2.A.Equals(t1.B) && !t2.A.Equals(t1.C) {
```
`switch`

Bastiantheone replied:
done

vladimir-ch commented:
If possible, I will try to take a look at this over the weekend. Before I start, could you give me a big picture about:

* What kind of algorithm this implements, for example, is there a book or a reference implementation that you follow?
* How do you handle the geometric predicates? From a quick look it seems to me that a plain floating point implementation is used which will not work reliably.
* It already called @kortschak 's attention, I'm also not convinced about the need to pass an index/id with each point.

Bastiantheone commented:
- The insertion of points is done with a hierarchical insertion algorithm based on this paper: [http://page.mi.fu-berlin.de/faniry/files/faniry_aims.pdf](http://page.mi.fu-berlin.de/faniry/files/faniry_aims.pdf). Basically there is a root triangle that contains all points. When a point is inserted, the triangle the point is in needs to be found. Starting at the root I look at the children and which one contains the point. Then I do the same thing for the triangle that contained the point and I keep doing that until I reach a leaf triangle. Then the point is inserted and forms three new triangles. These triangles are the children of the triangle that was just found. Then the edges are validated and flipped if they are not valid.
In the removeP function you'll notice a lot of code commented out. I was implementing the low degree optimization based on this paper : [https://hal.archives-ouvertes.fr/inria-00560379/document](https://hal.archives-ouvertes.fr/inria-00560379/document). It has problems when the points adjacent to the point to be removed form a non-convex polygon. The removal algorithm in base is triangulate and sew. The points adjacent to the point to be removed start a new triangulation and all triangle inside the polygon are kept.

- I have not implement anything yet. The only methods that can be affected are `triangle/inTriangle` and `point/Equals`. What would you suggest I should do?

- I need the ID for the way I implemented removal of points, when I start a new triangulation. Otherwise the ID is there for the user to make points easy to identify. I will need an ID or a pointer to implement the clustersequence. So I thought, I can just use the ID for two things at once. The comments on the fifth suggestion on fastjet/internal/delaunay/point.go by kortschak have more information on the removal part.

kortschak commented:
@vladimir-ch the discussion is [here](https://github.com/go-hep/hep/pull/72#discussion_r125801069).

An alternative to the `ID` field approach for deletion is to use a pointer chain and add a `Root() *Point` method. When a point is deleted the new points to the parent that was deleted. This leaves the user details of IDs and so on completely agnostic to use, the client can make a `map[*Point]int` to store `int` IDs if they need. Now this may seem more complicated, but at the moment, the requirements for ID are inadequately defined in the documentation, and that makes me feel queasy. If the requirements were more clearly stated in the docs, then maybe I'd feel better.

What the pointer chain approach does not handle is the use in `inPolygon` because they can be arbitrary words, but as the documentation stands (while I understand the basis for the test) I don't understand how an invariant-free ID fits with their use in that context.

As is always the case documentation always helps, a couple of examples of what I think are good documentation in unexported code are [here](https://github.com/gonum/gonum/blob/master/graph/community/louvain_directed.go#L409-L451) and [here](https://github.com/gonum/gonum/blob/master/graph/path/shortest.go#L110-L147). These are both mine, though the quality of the documentation can be put firmly at the feet of @vladimir-ch because of his excellent eye for documentation during code reviews.

Bastiantheone commented on fastjet/internal/delaunay/point.go:
```diff
@@ -0,0 +1,102 @@
+// looks at all adjacent points and finds the nearest one
+func (p *Point) findNearest() {
+	var np *Point
+	min := math.MaxFloat64
```
Done

Bastiantheone commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,193 @@
+	}
+}
+
+func clockwise(a *Point, b *Point, c *Point) bool {
```
Done

kortschak commented:
I will be unavailable for reviewing for a while. I will TAL when I can.

vladimir-ch commented:
I took a quick look but didn't have time to completely sort my thoughts on this.

Regarding the geometric predicates, they are used whenever you need to make decisions based on the geometry, that is, in places like `inTriangle` or when swapping an edge to make it Delaunay (but not in comparing points, the points are given to us and we don't modify them here). Doing it robustly is definitely not easy, and honestly I don't know how to write them from scratch. `math/big` could be of help. Anyway, I would delegate them to an interface and float64 could be one straghtforward implementation. The user then could choose among various offering different trade offs.

I will add further comments in the code.

vladimir-ch commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,663 @@
+
+// NewDelaunay creates a delaunay triangulation with the given points
+// all points have to be inside the user defined bounds
+func NewDelaunay(points []*Point, maxX float64, minX float64, maxY float64, minY float64) *Delaunay {
```
1.  With the Point interface this would take `points []Point.
2. I'm not convinced about the need of providing a bounding rectangle. What if the points are not known a priori? Currently the code panics if the rectangle is wrong. It seems to me that the rectangle exists only for building the tree for quickly locating the containing triangle. How does it compare to other approaches, for example "walking" on the mesh towards to point. This would need no bounding rectangle, and no additional data structure. If giving the bounding rectangle were significantly faster, providing it could be optional.
3. Currently `points` can be nil and providing them has no benefit. With some preprocessing assumptions could be made and the initial points could be inserted more efficiently. Otherwise `NewDelaunay` could be simplified to just:
```
t := delaunay.New()
for _, p := range points {
    t.InsertPoint(p)
}
```
with some mechanism for specifying an implementation of geometric predicates, perhaps the bounding rectangle (if considered useful) and any other settings.
4. The bounds should be given as (min, max).

Bastiantheone replied:
According to this [paper](https://hal.inria.fr/inria-00166711/document) the time complexity of the walk has a worst case of O(n^5/3). I can add the walk as an option to the user.

Bastiantheone replied:
@vladimir-ch are you familiar with the walking algorithm? I ran into a problem I can't seem to figure out trying to implement it. I asked the question in [stackoverflow](https://stackoverflow.com/questions/45065506/delaunay-walk-point-outside-triangles). Can you take a look and see if you can help

Bastiantheone replied:
Nevermind, I think I figured it out

vladimir-ch commented:
I didn't look at the removing of points from the triangulation and at voronoi.go
Also, "Triangulations and Applications" from Springer is a very accessible book on this topic.

Bastiantheone commented on fastjet/internal/delaunay/point.go:
```diff
@@ -0,0 +1,137 @@
+}
+
+// NewPoint returns Point for the given x,y coordinates and id
+func NewPoint(x float64, y float64, id int) *Point {
```
@vladimir-ch Can I just make the ID a private field and increment it for each point inserted? I'm not quite sure how you were thinking of implementing this with the interface variable p.

kortschak replied:
I too prefer the idea of having `Point` be an interface.

vladimir-ch commented on fastjet/internal/delaunay/point.go:
```diff
@@ -50,13 +51,21 @@ func (p *Point) inTriangle(t *Triangle) (inside, edge bool) {
 
-	// inside triangle
-	if barcen1 > 0 && barcen1 < 1 && barcen2 > 0 && barcen2 < 1 && barcen3 > 0 && barcen3 < 1 {
+	// use the math/big package to handle the geometric predicates
```
I don't think this will work. If anything, already the input point coordinates would have to be converted into arbitrary-precision numbers. Since the computations with them are slow, first we should try with float64 and if the result is not reliable, resort to big.

The predicates should be really factored out into its own type that satisfies an interface, and carefully and thoroughly tested (random point, co-linear points, almost co-linear points, four points on a circle, four points almost on a circle, etc). Then all geometry-based decisions should be done using these predicates. I think only two are necessary:
```
type OrientationKind int
const (
    CCW OrientationKind = iota
    CW
    Colinear
)

type Predicate interface {
    // Orientation returns how the point (x3,y3) is oriented with respect to
    // the line defined by the points (x1,y1) and (x2,y2).
    Orientation(x1, y1, x2, y2, x3, y3 float64) OrientationKind

    // InCircumcircle returns whether the point (x,y) lies inside
    // the circle defined by the points (x1,y1), (x2,y2), (x3,y3).
    // Maybe a more refined return value could be useful.
    InCircumcircle(x1, y1, x2, y2, x3, y3, x, y float64) bool
}
```
The predicates should be implemented so that eventually the decision boils down to comparing a single number to 0. In case of `Orientation` this would be checking the sign of the determinant
```
|x1 y1 1|
|x2 y2 1|
|x3 y3 1|
```
Here the decisions are complex and spread over many places.

I'm no expert on this and I'm not saying it has to be as I write above but the predicates are the basic building block, they must be robust, and unlike the rest of the algorithm they involve quite a bit of magic.

vladimir-ch replied:
Just an additional comment, `inTriangle` then could be implemented by calling `Orientation` three times, once for each triangle edge and the point being checked. If it returns CCW three times (assuming the triangles are oriented counterclockwise), the point is inside.

Bastiantheone replied:
implemented inTriangle using orientation. 

Where would I use OrientationKind in the code and did you make orientation and inCircumCircle public on purpose?

vladimir-ch replied:
My original idea was that there could be several implementations of the predicates with various levels of robustness vs speed tradeoff. It could be configurable and the user could even provide its own external implementation. This is of course not necessary and you can just implement these as two unexported functions.

kortschak commented on fastjet/internal/delaunay/delaunay_test.go:
```diff
@@ -0,0 +1,461 @@
+			t.Fatalf("Triangle T%s not as expected", ts[i])
+		}
+	}
+	expv := []*Point{NewPoint(-2.523, 2.25, -1),
```
Split these over lines.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/delaunay_test.go:
```diff
@@ -0,0 +1,461 @@
+	}
+}
+
+func BenchmarkNewDelaunay50(b *testing.B) {
```
If we are happy to require go1.7, then these can be in a for loop calling subtests rather than separate functions.

kortschak commented on fastjet/internal/delaunay/point.go:
```diff
@@ -0,0 +1,154 @@
+	adjacentTriangles triangles
+	// isOutside indicates whether the point is outside the triangulation.
+	isOutside bool
+	// ID is a user defined identifier. There are no restrictions on how the user uses it.
```
This still makes me very queasy. I don't like that a free user value is available that is used internally under strict invariants. Especially on a pointer value. This is how subtle bugs are born. Please find another way - at least discuss possible options that do not leave this open.

One approach is to keep ID only for internal use.

Bastiantheone replied:
For now, I changed to be used internally only

kortschak commented on fastjet/internal/delaunay/point.go:
```diff
@@ -0,0 +1,154 @@
+	// polygon formed by all those points.
+	ID      int
+	nearest *Point
+	// squared distance to the nearest neighbor
```
`// dist is the squared distance to the nearest neighbor.`

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/point.go:
```diff
@@ -0,0 +1,154 @@
+type Point struct {
+	// X and Y are the coordinates of the point.
+	X, Y float64
+	// adjacentTriangles is used for the triangulation, to find the nearest neighbor and
```
`// adjacentTriangles is a list of triangles containing the point.` is enough - correct this if I have misinterpreted it.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/point.go:
```diff
@@ -0,0 +1,154 @@
+	b1 := big.NewFloat(barcen1)
+	b2 := big.NewFloat(barcen2)
+	b3 := big.NewFloat(barcen3)
+	zero := big.NewFloat(0.0)
```
Make `zero` and `one` package-level vars to avoid allocation.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,190 @@
+			(t.C.Equals(s.A) || t.C.Equals(s.B) || t.C.Equals(s.C))
+}
+
+func (t *Triangle) String(name string) string {
```
Please make `*Triangle` a `fmt.Stringer`. The signature should be `String() string`.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,190 @@
+			x = (b2 - b1) / (m1 - m2)
+			y = m1*x + b1
+			if math.IsNaN(y) {
+				panic(fmt.Errorf("delaunay: error caluclating the circumcenter of triangle " + t.String("T")))
```
panic(fmt.Errorf("delaunay: error calculating the circumcenter of triangle %v", t))

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,190 @@
+
+// appendT appends to a slice of triangles and updates the nearest neighbor
+// it is used when the adjacent triangles of a point change
+func (triangles triangles) appendT(elems ...*Triangle) []*Triangle {
```
Please make the receiver simply `t`. `elems` could be renamed `triangles` since the type is already dealt with.

`s/appendT/append/g`; the signature tells people what is being appended.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,190 @@
+func (triangles triangles) appendT(elems ...*Triangle) []*Triangle {
+	// check if nearest neighbor changes by going through each triangles points
+	// and checking if the distance to that point is less. It is done both ways.
+	for _, t := range elems {
```
`s/t/e/g` since the receiver will be `t`, or if you rename `elems`, `tri`. This should also be applied below to `remove` and `finalize`.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// remove removes given triangles from a slice of triangles
+func (triangles triangles) remove(elems ...*Triangle) []*Triangle {
```
Make the receiver `t`.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,190 @@
+	// check if nearest neighbor of any point is removed and if so
+	// put that point in the update slice
+	var update []*Point
+	for _, t := range elems {
```
`s/t/e/g`

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,190 @@
+		}
+	}
+	for i := len(triangles) - 1; i >= 0; i-- {
+		for j, tri := range elems {
```
`s/tri/e/g`

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,190 @@
+// finalize returns the final delaunay triangles
+// only keeps leaf elements from the hierarchy
+// removes given triangles
+func (triangles triangles) finalize(elems ...*Triangle) []*Triangle {
```
Receiver to be `t`.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+// all points have to be inside the user defined bounds
+// uses a hierarchy to find the triangle which contains a point
+// has a worst time complexity of O(nln(n))
+func HierarchicalDelaunay(points []*Point, maxX float64, minX float64, maxY float64, minY float64) *Delaunay {
```
Why was this renamed? `NewDelaunay` works fine for me.

Also note that the params can be `(pts []*Point, maxX, minX, maxY, minY float64) *Delaunay`.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+	return d
+}
+
+// WalkDelaunay creates a delaunay triangulation with the given points
```
Rename this to `NewUnboundedDelaunay`. This gives a better indication of the intention.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+	if len(points) < 3 {
+		panic(fmt.Errorf("delaunay: not enough points"))
+	}
+	rand.Seed(time.Now().UnixNano())
```
Take a `*rand.Rand` parameter so that the user can decide what the seed is. If it is `nil` use `rand.Intn`, but a library should not contaminate global variable behind a user's back.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+
+// Triangle returns all delaunay Triangles
+func (d *Delaunay) Triangles() []*Triangle {
+	if d.hierarchy {
```
Invert this logic to reduce overall indentation.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+func (d *Delaunay) InsertPoint(p *Point) {
+	p.adjacentTriangles = make(triangles, 0)
+	var t *Triangle
+	var onE bool
```
`isOnEdge`?

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+	previous1, previous2 := r, l
+	outer1 = append(outer1, l)
+	outer2 = append(outer2, r)
+	found1, found2 := true, true
```
Move `outer2` and `found2 to where they are first used.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+	outer1 = append(outer1, l)
+	outer2 = append(outer2, r)
+	found1, found2 := true, true
+	for i := 0; found1; i++ {
```
It looks to me like this logic can be factored out into a helper for the two points.

Bastiantheone replied:
One is clockwise and the other one is counterclockwise. Should I use a helper anyways?

kortschak replied:
If you can without too much acrobatics.

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+	root                   *Triangle
+	maxX, minX, maxY, minY float64
+	// hierarchy indicates which method to use to find the triangle that contains the point
+	hierarchy bool
```
`s/hierarchy/useHierarchical/g`

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+	l.adjacentTriangles = l.adjacentTriangles.appendT(nts[0])
+	r.adjacentTriangles = r.adjacentTriangles.appendT(nts[0])
+	for i := 0; i < len(outer1)-1; i++ {
+		nt := NewTriangle(outer1[i+1], p, outer1[i])
```
This part can also be in the helper.

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+		p2 = t.A
+		p3 = t.B
+	default:
+		panic(fmt.Errorf("fastjet: delaunay can't find point P%v in Triangle T%v", p, t))
```
`s/fastjet/`?`/`

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+	d.validateEdge(nt4, p)
+}
+
+// validateEdge turns triangle into a valid delaunay triangle
```
The use of "validate" here is difficult. The term is usually used to mean something similar to the C++ assert. That is not what you mean though, according to the comment (and with some reading, the code - the actual mutation occurs in `flip`).

Bastiantheone replied:
How is `makeEdgeDelaunay`?

kortschak replied:
`make` (and `new`) have meaning in Go, they usually return a value. This doesn't do that so I would avoid that name.

`func (d *Delaunay) addTriangleOpposite(p *Point, t *Triangle) {`?

Bastiantheone replied:
I don't think it reflects well on the purpose of the function. Is `legalizeEdge` better?

kortschak replied:
No, I don't like that. If the name I suggested is incorrect, find another, but "legalize" doesn't tell the reader what the code is intended to do.

I addition to the Go documentation style, it would be worth reading [Pike Style](https://www.lysator.liu.se/c/pikestyle.html) and [Russ' post on naming things](https://research.swtch.com/names) and its comments.

I'm going to be unavailable from the end of the week (semester starts and I am teaching), but overall, I think the documentation still needs improvement (this is more important than the actual code at this stage). The single biggest impediment to reviewing this I have is that it's not always immediately clear what any particular bit of code is intended to do.

To address the function naming issue here, write good documentation for the methods, and then make a meaningful precis of that become the name. The issues with `legalizeEdge` for me are two-fold: 1) "legalise" could be either make something match what it legal, or allow what is not previously legal - this is presumably the former, but need not necessarily be; and 2) there is no explicit edge in the signature, so the name makes me have to think to understand what it being done. Some thinking requirement is OK, but try to minimise the cognitive load on the reader (it may be you later).

Bastiantheone replied:
Okay, thank you. 

I have now changed the name to `findTriangleOpposite`

kortschak replied:
Can you explain why this is different to `addTriangleOpposite` in a way that is more accurate?

vladimir-ch replied:
The key operation this does is to swap the edge in t opposite to p so as to make it locally optimal or Delaunay. For me the clearest name would be `flipDelaunay` (because there is the unconditional `flip` function) or `swapDelaunay` (because I'm used to swap rather than flip terminology).

Bastiantheone replied:
I like that @vladimir-ch

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+	}
+}
+
+// flip flips edge between two triangles
```
Can you clarify what to "flip" an edge between two triangles means.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+	case !t2.C.Equals(t1.A) && !t2.C.Equals(t1.B) && !t2.C.Equals(t1.C):
+		p2 = t2.C
+	default:
+		panic(fmt.Errorf("delaunay: triangle T2%v doesn't have points not in T1%v", t2, t1))
```
Please don't use double negatives.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+// flip flips edge between two triangles
+func (d *Delaunay) flip(t1 *Triangle, t2 *Triangle) (*Triangle, *Triangle) {
+	// find points adjacent and opposite to edge
+	var p1, p2, pO1, pO2 *Point
```
`var adj1, adj2, opp1, opp2 *Point`

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+}
+
+// flip flips edge between two triangles
+func (d *Delaunay) flip(t1 *Triangle, t2 *Triangle) (*Triangle, *Triangle) {
```
`func (d *Delaunay) flip(t1, t2 *Triangle) (nt1, nt2 *Triangle) {`

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,190 @@
+	children triangles
+	// A,B,C are the points that make up the triangle
+	A, B, C *Point
+	// inD holds whether a triangle is part of the triangulation
```
`s/inD/isInTriangulation/g`

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/voronoi.go:
```diff
@@ -0,0 +1,353 @@
+	"math"
+)
+
+// Voronoi holds the border information
```
This is implementation details, the user want to know how they use it.

Bastiantheone replied:
Done

Bastiantheone commented:
Here are the bench test result for the walk
```
300	   5050283 ns/op	  199376 B/op	    7146 allocs/op
100	  13170729 ns/op	  501286 B/op	   17637 allocs/op
50	  22401214 ns/op	  840157 B/op	   29308 allocs/op
50	  31881838 ns/op	 1197957 B/op	   41466 allocs/op
30	  42502483 ns/op	 1595459 B/op	   54833 allocs/op
30	  54203010 ns/op	 2003058 B/op	   68345 allocs/op
20	  67053940 ns/op	 2425061 B/op	   82676 allocs/op
20	  77354280 ns/op	 2857705 B/op	   96985 allocs/op
20	  90005110 ns/op	 3320040 B/op	  112365 allocs/op
10	 100707680 ns/op	 3775046 B/op	  127839 allocs/op
10	 114304270 ns/op	 4237028 B/op	  143062 allocs/op
10	 128609020 ns/op	 4718756 B/op	  158748 allocs/op
10	 141805900 ns/op	 5201448 B/op	  175565 allocs/op
10	 157866490 ns/op	 5725302 B/op	  192481 allocs/op
10	 172009610 ns/op	 6288319 B/op	  211360 allocs/op
10	 184712550 ns/op	 6759535 B/op	  227356 allocs/op
10	 198378770 ns/op	 7301284 B/op	  243758 allocs/op
5	 218212680 ns/op	 7897428 B/op	  263961 allocs/op
5	 231416960 ns/op	 8401136 B/op	  280996 allocs/op
5	 245613780 ns/op	 8940987 B/op	  298926 allocs/op
```

Bastiantheone commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+	return d.triangles.finalize()
+}
+
+func (d *Delaunay) InsertPoint(p *Point) {
```
Done

Bastiantheone commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1066 @@
+	}
+}
+
+func (d *Delaunay) RemovePoint(p *Point) {
```
Done

Bastiantheone commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -142,12 +332,128 @@ func (d *Delaunay) RemovePoint(p *Point) {
 
+// removeOuter finds the point around a point to be removed in clockwise order.
+// it is only used by the walk method
+// ind is the index of the last point found in the points slice
```
Done

Bastiantheone commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -142,12 +332,128 @@ func (d *Delaunay) RemovePoint(p *Point) {
+			i++
+		}
+	}
+	// call removeP with the new found polygon
```
Done

Bastiantheone commented on fastjet/internal/delaunay/delaunay_test.go:
```diff
@@ -0,0 +1,461 @@
+
+const tol = 1e-3
+
+func TestSimple(t *testing.T) {
```
Done

Bastiantheone commented on fastjet/internal/delaunay/delaunay_test.go:
```diff
@@ -0,0 +1,461 @@
+	}
+	nn := make([]*Point, 0, len(points)-1)
+	nd := make([]float64, 0, len(points)-1)
+	for i, p := range points {
```
Done

Bastiantheone commented:
I'm not quite sure why this build failed @sbinet

kortschak commented on fastjet/internal/delaunay/triangle.go:
```diff
@@ -0,0 +1,190 @@
+
+// appendT appends to a slice of triangles and updates the nearest neighbor
+// it is used when the adjacent triangles of a point change
+func (triangles triangles) append(elems ...*Triangle) []*Triangle {
```
This should be `func (t triangles) append(triangles ...*Triangle) []*Triangle {` or something similar. In Go terminology, the receiver is the type on which the method is implemented.

kortschak commented:
Rather than your second fix (which has a lot of moving parts), I would do this (passes):

```diff
diff --git a/fastjet/internal/delaunay/triangle.go b/fastjet/internal/delaunay/triangle.go
index 605527d..0b7b848 100644
--- a/fastjet/internal/delaunay/triangle.go
+++ b/fastjet/internal/delaunay/triangle.go
@@ -99,61 +99,61 @@ type triangles []*Triangle
 
 // appendT appends to a slice of triangles and updates the nearest neighbor
 // it is used when the adjacent triangles of a point change
-func (triangles triangles) append(triangles ...*Triangle) []*Triangle {
+func (t triangles) append(triangles ...*Triangle) []*Triangle {
        // check if nearest neighbor changes by going through each triangles points
        // and checking if the distance to that point is less. It is done both ways.
-       for _, t := range triangles {
-               d := t.A.distance(t.B)
-               if d < t.A.dist {
-                       t.A.dist = d
-                       t.A.nearest = t.B
+       for _, tri := range triangles {
+               d := tri.A.distance(tri.B)
+               if d < tri.A.dist {
+                       tri.A.dist = d
+                       tri.A.nearest = tri.B
                }
-               if d < t.B.dist {
-                       t.B.dist = d
-                       t.B.nearest = t.A
+               if d < tri.B.dist {
+                       tri.B.dist = d
+                       tri.B.nearest = tri.A
                }
-               d = t.B.distance(t.C)
-               if d < t.B.dist {
-                       t.B.dist = d
-                       t.B.nearest = t.C
+               d = tri.B.distance(tri.C)
+               if d < tri.B.dist {
+                       tri.B.dist = d
+                       tri.B.nearest = tri.C
                }
-               if d < t.C.dist {
-                       t.C.dist = d
-                       t.C.nearest = t.B
+               if d < tri.C.dist {
+                       tri.C.dist = d
+                       tri.C.nearest = tri.B
                }
-               d = t.A.distance(t.C)
-               if d < t.A.dist {
-                       t.A.dist = d
-                       t.A.nearest = t.C
+               d = tri.A.distance(tri.C)
+               if d < tri.A.dist {
+                       tri.A.dist = d
+                       tri.A.nearest = tri.C
                }
-               if d < t.C.dist {
-                       t.C.dist = d
-                       t.C.nearest = t.A
+               if d < tri.C.dist {
+                       tri.C.dist = d
+                       tri.C.nearest = tri.A
                }
        }
-       return append(triangles, triangles...)
+       return append(t, triangles...)
 }
 
 // remove removes given triangles from a slice of triangles
-func (triangles triangles) remove(triangles ...*Triangle) []*Triangle {
+func (t triangles) remove(triangles ...*Triangle) []*Triangle {
        // check if nearest neighbor of any point is removed and if so
        // put that point in the update slice
        var update []*Point
-       for _, t := range triangles {
-               if t.A.nearest != nil && (t.A.nearest.Equals(t.B) || t.A.nearest.Equals(t.C)) {
-                       update = append(update, t.A)
+       for _, tri := range triangles {
+               if tri.A.nearest != nil && (tri.A.nearest.Equals(tri.B) || tri.A.nearest.Equals(tri.C)) {
+                       update = append(update, tri.A)
                }
-               if t.B.nearest != nil && (t.B.nearest.Equals(t.A) || t.B.nearest.Equals(t.C)) {
-                       update = append(update, t.B)
+               if tri.B.nearest != nil && (tri.B.nearest.Equals(tri.A) || tri.B.nearest.Equals(tri.C)) {
+                       update = append(update, tri.B)
                }
-               if t.C.nearest != nil && (t.C.nearest.Equals(t.A) || t.C.nearest.Equals(t.B)) {
-                       update = append(update, t.C)
+               if tri.C.nearest != nil && (tri.C.nearest.Equals(tri.A) || tri.C.nearest.Equals(tri.B)) {
+                       update = append(update, tri.C)
                }
        }
-       for i := len(triangles) - 1; i >= 0; i-- {
-               for j, t := range triangles {
-                       if t.Equals(triangles[i]) {
-                               triangles = append(triangles[:i], triangles[i+1:]...)
+       for i := len(t) - 1; i >= 0; i-- {
+               for j, tri := range triangles {
+                       if tri.Equals(t[i]) {
+                               t = append(t[:i], t[i+1:]...)
                                triangles = append(triangles[:j], triangles[j+1:]...)
                                break
                        }
@@ -163,28 +163,28 @@ func (triangles triangles) remove(triangles ...*Triangle) []*Triangle {
        for _, p := range update {
                p.findNearest()
        }
-       return triangles
+       return t
 }
 
 // finalize returns the final delaunay triangles
 // only keeps leaf elements from the hierarchy
 // removes given triangles
-func (triangles triangles) finalize(triangles ...*Triangle) []*Triangle {
-       ft := make([]*Triangle, 0, len(triangles))
-       for _, t := range triangles {
-               if t.isInTriangulation {
+func (t triangles) finalize(triangles ...*Triangle) []*Triangle {
+       final := make([]*Triangle, 0, len(triangles))
+       for _, a := range t {
+               if a.isInTriangulation {
                        keep := true
-                       for j, tri := range triangles {
-                               if tri.Equals(t) {
+                       for j, b := range triangles {
+                               if b.Equals(a) {
                                        keep = false
                                        triangles = append(triangles[:j], triangles[j+1:]...)
                                        break
                                }
                        }
                        if keep {
-                               ft = append(ft, t)
+                               final = append(final, a)
                        }
                }
        }
-       return ft
+       return final
 }
```

Note that renaming many things at once is intrinsically fraught, and should probably usually go through an intermediate stage where no collision is possible.

Bastiantheone commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -142,12 +332,128 @@ func (d *Delaunay) RemovePoint(p *Point) {
+// removeOuter finds the point around a point to be removed in clockwise order.
+// it is only used by the walk method
+// ind is the index of the last point found in the points slice
+func (d *Delaunay) removeOuter(p *Point, points []*Point, ind int) {
```
Done

Bastiantheone commented on fastjet/internal/delaunay/delaunay_test.go:
```diff
@@ -0,0 +1,461 @@
+	b := NewPoint(0, 2, 1)
+	c := NewPoint(1, 0, 2)
+	d := NewPoint(4, 4, 3)
+	ep := NewPoint(3, 2, -1)
```
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1162 @@
+)
+
+var (
+	Zero = big.NewFloat(0)
```
Please don't make these exported.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1162 @@
+	ab := big.NewFloat(p.orientation(t.A, t.B))
+	bc := big.NewFloat(p.orientation(t.B, t.C))
+	ca := big.NewFloat(p.orientation(t.C, t.A))
+	zero := big.NewFloat(0)
```
Delete this.

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1162 @@
+}
+
+// insertPonE inserts a point on an edge
+func (d *Delaunay) insertPonE(t1 *Triangle, p *Point) {
```
`func (d *Delaunay) insertAtEdge(new *Point, t *Triangle) {`

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1162 @@
+	d.validateEdge(t3, p)
+}
+
+func (d *Delaunay) insertPonBorderE(t *Triangle, p *Point) {
```
This seems very similar to `insertPonE`, there needs to be documentation explaining what these do.

`func (d *Delaunay) insertAtBorderEdge(new *Point, t *Triangle) {`

Bastiantheone replied:
Done

kortschak commented on fastjet/internal/delaunay/triangulation.go:
```diff
@@ -0,0 +1,1162 @@
+}
+
+// walkTriangle finds the triangle which contains p by using a remembering stochastic walk
+func (d *Delaunay) walkTriangle(t *Triangle, p *Point) (*Triangle, bool, *Point, *Point) {
```
The returns should have names here.

Bastiantheone replied:
Done

vladimir-ch commented:
This test added to `delaunay_test.go` panics for `n >= 5`:
```
func grid(nx, ny int, angle float64) []*Point {
	s := math.Sin(angle)
	c := math.Cos(angle)
	var points []*Point
	for xi := 0; xi < nx; xi++ {
		tx := float64(xi)
		for yi := 0; yi < ny; yi++ {
			ty := float64(yi)
			x := tx*c - ty*s
			y := tx*s + ty*c
			points = append(points, NewPoint(x, y))
		}
	}
	return points
}

func TestDelaunayGrid(t *testing.T) {
	const degrees = math.Pi / 180
	const n = 5
	p := grid(n, n, 10*degrees)
	NewUnboundedDelaunay(p, nil)
}
```

Bastiantheone commented:
The issue is fixed. I needed to change the inTriangle function like you suggested @vladimir-ch

Bastiantheone commented:
The problem with the grid test is that some triangles start overlapping. It has most likely something do to with floating point imprecision. Right now my guess is  that the circumcircle test is the root cause. Did I use math/big the right way in commit e6f0dde @vladimir-ch ? And should I do it similar in inCircumcircle and centerOfCircumcircle?

vladimir-ch commented:
Having predicates that work robustly even in these degenerate cases and then handling all the possible cases is not easy. There are no triangles that overlap, the points are well separated and the edges between them are unique, so it's probably the predicates that return inconsistent results for neighbouring triangles caused by floating point arithmetic.

As I wrote, I'm no expert on writing robust predicates. If I were to write them anyway, I would definitely base them on the sign of a corresponding determinant. A simple approach could be: If the determinant is "large", floating point computation is reliable (and fast). If it is "small", I would convert the coordinates to big.Rat (the conversion is exact) and perform the computation again with (much slower but exact) big.Rat. However, I don't know how to reliably judge what is small and large. For example, a very long skinny triangle can be almost degenerate, but it's area might not be small in absolute sense.

This PR is too big for me to provide a useful feedback. I would suggest to close this (unless @kortschak objects) and resubmit it in a series of smaller PRs in the following order:
1. Robust geometric predicates with careful tests
2. _Fast_ triangulation data structure that provides all the necessary operations
3. An algorithm for creating the Delaunay triangulation built on top of 1. and 2.
4. Only then add extras like removal of points, Voronoi, ...

Even each step here should be split into several PRs.

For a future reference for point 3, I'm still not convinced that providing an a priori bounding rectangle/triangle is necessary even for the hierarchical algorithm. It should be possible to work around this having a dummy triangle internally that encloses everything.

Bastiantheone commented:
About the dummy triangle. I wanted to make sure that the triangle encloses everything and is not too close to the points, because it will change up the triangulation, that's why I asked for the user input. So am I understanding it correctly, that you suggest to just use a very large triangle as root triangle?

kortschak commented:
>This PR is too big for me to provide a useful feedback. I would suggest to close this (unless @kortschak objects)

I have absolutely no objection to this.

Bastiantheone commented:
I started working on the predicate implementation. If accurate results are required, then the user coordinates need to be accurate too. Would it make sense to let the user pick between accurate results and approximations and for the accurate results the user would have to give the coordinates as big/rat values, meaning that eta and phi need to be accurate as well @go-hep/developers ? To me it doesn't make sense accurately calculating the determinant with inaccurate coordinates.

vladimir-ch commented:
> ... it doesn't make sense accurately calculating the determinant with inaccurate coordinates.

The purpose of the predicates is to be able to make decisions based on geometry. You can implement Delaunay triangulation only once you can rely on what Orientation and InCircle return. Floating point does not provide that for (almost) degenerate cases so you need something extra.  This has nothing to do with the input coordinates being accurate or not from the user application viewpoint.

As an example, for `Orientation` I would try something like this (untested):
```
type RationalPredicates struct{}

func (RationalPredicates) Orientation(x1, y1, x2, y2, x3, y3 float64) OrientationKind {
	// Compute the determinant of the matrix
	//  | x1 y1 1 |
	//  | x2 y2 1 |
	//  | x3 y3 1 |
	// det := x1*y2 + x2*y3 + x3*y1 - x1*y3 - x2*y1 - x3*y2
	rx1 := new(big.Rat).SetFloat64(x1)
	ry1 := new(big.Rat).SetFloat64(y1)
	rx2 := new(big.Rat).SetFloat64(x2)
	ry2 := new(big.Rat).SetFloat64(y2)
	rx3 := new(big.Rat).SetFloat64(x3)
	ry3 := new(big.Rat).SetFloat64(y3)
	rdet := new(big.Rat).Mul(rx1, ry2)
	t := new(big.Rat)
	t.Mul(rx2, ry3)
	rdet.Add(rdet, t)
	t.Mul(rx3, ry1)
	rdet.Add(rdet, t)
	t.Mul(rx1, ry3)
	rdet.Sub(rdet, t)
	t.Mul(rx2, ry1)
	rdet.Sub(rdet, t)
	t.Mul(rx3, ry2)
	rdet.Sub(rdet, t)
	switch rdet.Sign() {
	case 1:
		return CCW
	case -1:
		return CW
	default:
		return Colinear
	}
}
```
This will definitely not be fastest but should give correct results.

Bastiantheone commented:
PR for the predicates: #75

sbinet commented:
I went thru all the email discussion (the web UI of github is a bit lacking when one comes after the battle).

I too would be inclined to split this PR into more manageable piece(s).
I see this has already somewhat started with go-hep/hep#75.
